api/v1beta1/openstackbaremetalset.go

```diff
@@ -190,19 +246,81 @@ func VerifyBaremetalSetScaleDown(
 	return nil
 }
 
-func verifyBaremetalSetInstanceLabelMatch(
-	l logr.Logger,
-	instance *OpenStackBaremetalSet,
-	bmh *metal3v1.BareMetalHost) (string, bool) {
+// Function to find valid assignments for computes-to-BMHs, given their labels, using backtracking
+func findValidBaremetalSetInstanceLabelAssignments(computes map[string]InstanceSpec, bmhs []metal3v1.BareMetalHost) map[string]metal3v1.BareMetalHost {
+	// First create map of valid computes-to-BMHs possibilities
+	computesToBmhs := map[string][]metal3v1.BareMetalHost{}
+	for compName, comp := range computes {
+		for _, bmh := range bmhs {
+			if IsMapSubset(bmh.GetLabels(), comp.BmhLabelSelector) {
+				computesToBmhs[compName] = append(computesToBmhs[compName], bmh)
+			}
+		}
+	}
+
+	// Now sort computes by the number of valid matches, as sorting helps in
```

Comment:
Do we need to all this? Can we not just sort the hosts in the `BaremetalHosts` and BMHs based on their labels and then we would always find the correct match?

Reply:
Do you think sorting like https://github.com/openstack-k8s-operators/openstack-baremetal-operator/compare/main...rabi:openstack-baremetal-operator:bmh_selection?expand=1  would work for the scenarios you're tesing? Should we add some functional testing for those scenarios?

Reply:
Yes.  I built your fork/branch and used it with the same series of manual tests I was trying, and everything worked as expected.  I'm not partial to my PR and would be just as comfortable with using your approach instead.  Basically I'm just interest it getting this working properly.

Here is an example of what happens currently.  I have an `OpenStackBaremetalSet` without any `bmhLabelSelector`s:

```
...
  baremetalHosts:
    edpm-compute-0: 
      ctlPlaneIP: 192.168.122.100
    edpm-compute-1: 
      ctlPlaneIP: 192.168.122.101
    edpm-compute-2:
      ctlPlaneIP: 192.168.122.102
  #bmhLabelSelector:
  #  app: openstack
...
```

When I go to create it, I see this:

```
$ oc apply -f baremetalset.yaml 
Error from server (Forbidden): error when creating "baremetalset.yaml": admission webhook "vopenstackbaremetalset.kb.io" denied the request: unable to find 3 requested BaremetalHosts in namespace openshift-machine-api for scale-up (0 in use, 2 available)

$ oc apply -f baremetalset.yaml 
Error from server (Forbidden): error when creating "baremetalset.yaml": admission webhook "vopenstackbaremetalset.kb.io" denied the request: unable to find 3 requested BaremetalHosts in namespace openshift-machine-api for scale-up (0 in use, 1 available)
```

You can see in the output `(0 in use, X available)` messages that the randomness of computes overwriting each other for BMH assignment is happening.  Anyhow, it seems like we're converging on a solution.